### PR TITLE
added backward function test for MPC Tensor

### PR DESCRIPTION
**Pull request by aanurraj**
## Description
Added backward tests for MPC test.

## Affected Dependencies
None

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 1]
```diff
@@ -472,3 +472,25 @@ def test_pow(get_clients, power) -> None:
...
+def test_backward(get_clients):
+    clients = get_clients(4)
+    session = Session(parties=clients)
+    session.autograd_active = True
+    SessionManager.setup_mpc(session)
+
+    x_secret = torch.tensor([[0.125, -1.25], [-4.25, 4], [-3, 3]], requires_grad=True)
+    y_secret = torch.tensor([[4.5, -2.5], [5, 2.25], [-3, 3]], requires_grad=True)
+    x = MPCTensor(secret=x_secret, session=session, requires_grad=True)
+    y = MPCTensor(secret=y_secret, session=session, requires_grad=True)
```
Comment (Syzygianinfern0) [inline tests/sympc/tensor/mpc_tensor_test.py:486]: Q: If you set `requires_grad=False`, does the `.backward()` not compute anything? 

Maybe: If so, we could have a test to check that...

Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py:486]: no :))
I was not able to write the test (and was  debugging if it was a issue or I am missing something)

Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py:486]: basically, if we don't set this to true, `Requires Grad` will be False and `GradFunc` will not be set. So it like calling backward on a normal tensor that has `requires_grad = False.`
Screenshot 2021-06-09 at 10 51 00 PM

Comment (gmuraru) [reply tests/sympc/tensor/mpc_tensor_test.py:486]: I think @Syzygianinfern0 wanted to validate that we do not have the gradient in that case.
Also, a test case when only one of the parameters has `requires_grad might be good.

x = MPC1 (without `requires grad`)
y = MPC2 (without `requires_grad`)

x - y (check if this one has `requires_grad` and check if `grad` is `None`

**Review by gmuraru — COMMENTED** [review]
Check comments

**Review by Syzygianinfern0 — APPROVED** [review]

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 2]
```diff
@@ -472,3 +472,45 @@ def test_pow(get_clients, power) -> None:
...
+    x = MPCTensor(secret=x_secret, session=session)
+    y = MPCTensor(secret=y_secret, session=session)
+
+    res_mpc = x - y
+    s_mpc = res_mpc.sum()
+    s_mpc.backward()
+
+    assert not res_mpc.requires_grad
+    assert x.grad is None
+    assert x.grad is None
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:516]: y grad

**Review by gmuraru — CHANGES_REQUESTED** [review]
Also, add a test where only one of the parameters has `requires_grad` (x has `requires_grad` and `y` not)

**Comment by Syzygianinfern0** [comment]
@gmuraru ping 🎯

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 3]
```diff
@@ -472,3 +472,68 @@ def test_pow(get_clients, power) -> None:
...
+
+    res_mpc = x - y
+    res = x_secret - y_secret
+    s_mpc = res_mpc.sum()
+    s = torch.sum(res)
+    s_mpc.backward()
+    s.backward()
+
+    assert res_mpc.requires_grad
+    assert np.allclose(x.grad.get(), x_secret.grad, rtol=1e-3)
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:538]: Q: Why not add the check for also `res.grad` and `res_mpc.grad`?

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 4]
```diff
@@ -472,3 +472,68 @@ def test_pow(get_clients, power) -> None:
...
+    x_secret = torch.tensor([[0.125, -1.25], [-4.25, 4], [-3, 3]])
+    y_secret = torch.tensor([[4.5, -2.5], [5, 2.25], [-3, 3]])
+    x = MPCTensor(secret=x_secret, session=session)
+    y = MPCTensor(secret=y_secret, session=session)
+
+    res_mpc = x - y
+    s_mpc = res_mpc.sum()
+    s_mpc.backward()
+
+    assert not res_mpc.requires_grad
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:514]: Q: Also here (I missed this last time), check that also `res_mpc.grad` is `None`?

**Comment by gmuraru** [comment]
> @gmuraru ping

Done - left some comments.

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 4 above)
Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py:514]: done

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 3 above)
Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py:540]: added an issue for this

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 2 above)
Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py]: done

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 5]
```diff
@@ -472,3 +472,70 @@ def test_pow(get_clients, power) -> None:
...
+    y = MPCTensor(secret=y_secret, session=session)
+
+    res_mpc = x - y
+    res = x_secret - y_secret
+    s_mpc = res_mpc.sum()
+    s = torch.sum(res)
+    s_mpc.backward()
+    s.backward()
+
+    # TODO: add assert for res_mpc.grad and res.grad
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:538]: This needs to be addressed, right?

Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py:538]: Opened an issue for this!

**Review by gmuraru — APPROVED** [review]
LGTM!

**Review by Syzygianinfern0 — APPROVED** [review]
Merge?
